doc/specs/#182 - Support for installation of portable standalone apps.md

```diff
@@ -0,0 +1,190 @@
+---
+author: Demitrius Nelon @denelon
+created on: 2022-03-09
+last updated: 2022-03-09
+issue id: 182
+---
+
+# Portable/Standalone applications
+
+"For [#182](https://github.com/microsoft/winget-cli/issues/182)"
+
+## Abstract
+
+Several packages do not have an installer. They are simply a binary executable. This specification describes how these types of programs will be treated as "packages" by the Windows Package Manager.
+
+## Inspiration
+
+Just because an installer doesn't exist for a particular package doesn't mean you shouldn't be able to install it with the Windows Package Manager.
+
+## Solution Design
+
+The installer type “portable” will be added to the enumerated list of installer types to support these types of packages.
+
+When one of these packages is encountered, there are several aspects to “installing” these packages and supporting the “upgrade” behavior. Notable concerns include the location of the package, creating or updating an entry in Windows Apps & Features.
+
+### Command Line Arguments
+
+Users should be able to specify the location for where the program is “installed” on their machine with the “--location” argument.
+
+Some packages like GitLab Runner have a file name including extra metadata (gitlab-runner-windows-386.exe). The installation instructions suggest renaming the file after it has been downloaded. In that vein, portable / standalone executables should have their “command” value specified so the Windows Package Manager can perform a default rename behavior. In addition, a “--rename” argument should be added so the user can choose a value they prefer. To have the upgrade scenario honor the custom name, this information should be recorded in the installed packages data store.
+
+Some portable applications generate or consume other files. An additional argument "--purge" will be added for the uninstall scenario to remove all files and subsequently the directory if the user wishes. The corollary argument "--preserve" will be used to preserve files if the default setting has been modified.
+
+### Settings
+
+Users should have settings to be able to specify a new default location other than the system default for the Windows Package Manager.
+
+The default path for installing these packages is "%LOCALAPPDATA%/WinGet_Packages/User" for user based installs. The default path for installing these packages is "%LOCALAPPDATA%/WinGet_Packages/Machine"for machine wide installs.
```

Comment:
```suggestion
The default path for installing these packages is "%LOCALAPPDATA%/WinGet_Packages/User" for user based installs. The default path for installing these packages is "%PROGRAMDATA%/WinGet_Packages/Machine" for machine wide installs.
```
The localappdata for a specific user may not be accessible by all users on a machine, so I would recommend these be put into a central location. Perhaps they could all be put into programdata and the registry entries determine whether it is installed for all users or a single user?

Reply:
@ryfu-msft thoughts?

Reply:
I think that Program Files is the correct location for machine wide installs, but that is more my intuition than having looked at the current guidance.  I agree that the specific user's local app data is not the right location.

Reply:
If we go with Program Files, then Architecture should also be respected to place them in `Program Files` of `Program Files (x86)` respectively

Reply:
If a portable app expects to be able to write to the directory it is installed in (as many do, since they expect that they are installed per user), putting them in Program Files would probably break stuff since non adminstrators can't write there. 

This may open up a security can of worms, but what if when you installed a portable app for the system winget just put it in everyone's %appdata%? You'd be making several copies of it, but it would improve compatibility. (I'm thinking how the Teams Machine-Wide Installer works.) Alternatively, maybe symlinking the individual files the portable app comes with from Program Files to somewhere the user can write to (so they get the application itself from program files, but any new preferences or other data would be written to a folder they had access to).

Reply:
I feel that the expectation with a system installation is that all users (even those that have not logged in yet) will have access.  That makes it hard to do anything that involves updating every user's files (as well as being questionable in terms of respecting other users' ownership).  I think that it is also a security problem to allow any user to write to the shared location.

I don't think that we are trying to make every package work in every scenario; and this seems like an area where if the portable needs to write to its current location, it might just not work for system install and non-admin use cases.